fix(api-stream): Add throwOnError support in undici.stream

assalielmehdi opened the pull request:
[#1753](https://github.com/nodejs/undici/issues/1753)

mdashlw commented:
This doesn't handle body properly, see differences between `api-request.js` and `api-stream.js` in `onData` and `onComplete`
```js
await undici.request('https://api.github.com/notfound', {
  throwOnError: true
})
```
```
ResponseStatusCodeError: Response status code 403: Forbidden {
  code: 'UND_ERR_RESPONSE_STATUS_CODE',
  body: '\r\n' +
    'Request forbidden by administrative rules. Please make sure your request has a User-Agent header (https://docs.github.com/en/rest/overview/resources-in-the-rest-api#user-agent-required). Check https://developer.github.com for other possible causes.\r\n',
  status: 403,
  statusCode: 403,
  headers: {
    'cache-control': 'no-cache',
    'content-type': 'text/html; charset=utf-8',
    'strict-transport-security': 'max-age=31536000',
    'x-content-type-options': 'nosniff',
    'x-frame-options': 'deny',
    'x-xss-protection': '0',
    'content-security-policy': "default-src 'none'; style-src 'unsafe-inline'",
    connection: 'close'
  }
}
```
```js
await undici.stream(
  "https://api.github.com/notfound",
  { throwOnError: true },
  () => fs.createWriteStream("response.txt")
);
```
```
ResponseStatusCodeError: Response status code 403: Forbidden {
  code: 'UND_ERR_RESPONSE_STATUS_CODE',
  body: undefined,
  status: 403,
  statusCode: 403,
  headers: {
    'cache-control': 'no-cache',
    'content-type': 'text/html; charset=utf-8',
    'strict-transport-security': 'max-age=31536000',
    'x-content-type-options': 'nosniff',
    'x-frame-options': 'deny',
    'x-xss-protection': '0',
    'content-security-policy': "default-src 'none'; style-src 'unsafe-inline'",
    connection: 'close'
  }
}
```

mcollina reviewed (commented):
good work, can you add docs?

assalielmehdi commented:
Hi @mdashlw, thank you for your comment.
I think that it's pretty normal for the body to be `undefined` in the case of `undici.stream()` since the `HTTP` body is written to the `writable stream` created by the factory. However, I had to make sure that the `writable` in finished before throwing an error.
We can retrieve the body from a `PassThrough` opaque if needed (see unit test).

assalielmehdi commented:
Hi @mcollina, thank you 🙏🏼.
In docs, `options` is already of type [`RequestOptions`](https://undici.nodejs.org/#/docs/api/Dispatcher?id=parameter-requestoptions) that extends [`DispatchOptions`](https://undici.nodejs.org/#/docs/api/Dispatcher?id=parameter-dispatchoptions), and `throwOnError` is already explained in DispatchOptions.
I see nothing more to add in docs. WDYT?

mdashlw commented:
https://github.com/nodejs/undici/pull/1767/commits/8f7a6d79ff2d43af6997770a710e0086057e9ba3 kind of re-introduces the entire issue.
This will create `response.txt` with the actual response and still throw `ResponseStatusCodeError`.
```js
await undici.stream(
  "https://api.github.com",
  { throwOnError: true },
  () => fs.createWriteStream("response.txt")
);
```
```
ResponseStatusCodeError: Response status code 403: Forbidden {
  code: 'UND_ERR_RESPONSE_STATUS_CODE',
  body: undefined
  ...
}
```
```
response.txt
Request forbidden by administrative rules. Please make sure your request has a User-Agent header (https://docs.github.com/en/rest/overview/resources-in-the-rest-api#user-agent-required). Check https://developer.github.com for other possible causes.
```
IMO the most intuitive and convenient way would be to throw an error with special body handling, as in `.request()`, and not to call the factory.

mcollina commented on test/client-stream.js:
```diff
@@ -786,3 +786,42 @@ test('stream legacy needDrain', (t) => {
+        method: 'GET',
+        throwOnError: true,
+        opaque: passThrough
+      }, ({ opaque }) => opaque)
```
I would not expect this factory to be called at all.

assalielmehdi commented:
Since in `undici.request()`, the body is consumed before throwing the error, I figured we want the same from `undici.stream()`. I fixed the code to throw an error before even calling the factory, and updated the unit test accordingly.

@mdashlw for the body handling in `undici.request()`, the body is consumed before throwing, and we can't do so in `undici.stream()` unless we call the factory. So `body` will always be undefined in the error object.

mcollina commented:
You can consume the body without calling the factory.

assalielmehdi commented:
I updated the logic, to throw before calling the factory, and I handled the body so the thrown error contains it. I updated the tests accordingly. I hope that is the behavior we expect.

assalielmehdi commented:
Hi @mcollina, can you review the changes please? 😊

KhafraDev commented:
there are ci failures caused by a node-fetch test (`npm run test:node-fetch`)

ronag commented:
This will be fixed once we move `throwOnError` to an interceptor.

https://github.com/nodejs/undici/issues/2835
